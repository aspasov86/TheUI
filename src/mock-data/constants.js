import m17 from '../assets/M17.jpg'
import m18 from '../assets/M18.jpg'
import m39 from '../assets/M39.jpg'
import m43 from '../assets/M43.jpg'

export const trueLogin = {
    username: 'acko1986@yahoo.com',
    password: 'Cliveowen20.'
};

const ids = [9426446563, 7410887671, 9083503333, 9233201614];
const images = [m17, m18, m39, m43];

class Majstor {
    constructor(firstName, lastName, email, phoneNumber, stats, place, occupation) {
        this.id = (() => {
            let firstOne = ids[0];
            ids.splice(0, 1);
            return firstOne;
        })()
        this.image = (() => {
            let first = images[0];
            images.splice(0,1);
            return first;
        })()
        this.firstName = firstName
        this.lastName = lastName
        this.email = email
        this.phoneNumber = phoneNumber
        this.stats = stats
        this.place = place
        this.occupation = occupation
    };
};

export const majstors = [
    new Majstor('Jovan', 'Ilic', 'jova@ilic.com', '063/444-331', { brzina: 33, pedantnost: 51, cena: 56, ljubaznost: 47}, 'Zrenjanin', 'Vodoinstalater'),
    new Majstor('Milivoje', 'Zbulj', 'zbulj@mil.com', '061/5414-731', { brzina: 57, pedantnost: 71, cena: 16, ljubaznost: 87}, 'Jagodina', 'Elektricar'),
    new Majstor('Bojan', 'Gorkov', 'gore@yahoo.com', '069/4712-781', { brzina: 73, pedantnost: 51, cena: 96, ljubaznost: 87}, 'Beograd', 'Zidar'),
    new Majstor('Veroljub', 'Ciric', 'cira@gmail.com', '064/777-991', { brzina: 13, pedantnost: 21, cena: 46, ljubaznost: 67}, 'Novi Sad', 'Stolar'),
];