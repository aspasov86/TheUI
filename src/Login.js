import React, { Component } from 'react';
import {
    Button,
    Form,
    Grid,
    Segment,
    Message,
} from 'semantic-ui-react';
import { DataContext } from './DataProvider';
import { trueLogin } from './mock-data/constants'


class Login extends Component {
    state = {
        username: '',
        password: '',
        usernameError: false,
        passwordError: false,
        errorMessages: [],
    }

    usernameErrorMessage = 'Email must be properly formatted'
    passwordErrorMessage = 'Password must not be empty'

    changeHandler = (e) => {
        this.setState({
            [e.target.id]: e.target.value
        });
    }

    componentDidMount() {
        if (localStorage.getItem('e-majs-auth')) {
            this.goHome()
        }
    }

    goHome = () => {
        this.props.history.push('/overview')
    }

    resetComponent = () => {
        this.setState({
            username: '',
            password: '',
            usernameError: false,
            passwordError: false,
            errorMessage: [],
        })
    }

    isValid = () => {
        const { username, password } = this.state;
        const userCheck = username.indexOf('@') !== -1 && username.indexOf('.') !== -1;
        const passCheck = password !== '';
        let errorMessages = [];
        if (!userCheck) {
            errorMessages.push(this.usernameErrorMessage)
        }
        if (!passCheck) {
            errorMessages.push(this.passwordErrorMessage)
        }
        this.setState({ 
            usernameError: !userCheck, 
            passwordError: !passCheck,
            errorMessages,
        });
        if (userCheck && passCheck) {
            return true;
        } 
        return false;
    }

    renderErrorMessage = errorMessages => {
        const content = (
            <ul>
                {errorMessages.map((msg, i) => (
                    <li key={i} style={{textAlign: 'left'}}>{msg}</li>
                ))}
            </ul>
        )
        return  (
            <Message error header='Invalid Input' content={content}/>
        )
}

    submitHandler = context => {
        if (this.isValid()) {
            const { username, password } = this.state;
            if (username === trueLogin.username && password === trueLogin.password) {
                context(username, password);
                localStorage.setItem('e-majs-auth', username)
                this.goHome()
            } else if (username !== trueLogin.username) {
                const errorMessages = ['Email does not exist in our database'];
                this.setState({ usernameError: true, errorMessages });
            } else if (username === trueLogin.username && password !== trueLogin.password) {
                const errorMessages = ['You typed in a wrong password'];
                this.setState({ passwordError: true, errorMessages });
            } 
        }
    }

    render() {
        return (
            <div className='login-form'>
                <Grid textAlign='center' style={{ height: '100%' }} verticalAlign='middle'>
                    <Grid.Column style={{ maxWidth: 450, marginTop: '100px' }}>
                        <DataContext.Consumer>
                            {context => (
                                <Form 
                                    id="submit-btn" 
                                    size='large' 
                                    error={this.state.usernameError || this.state.passwordError} 
                                    onSubmit={() => this.submitHandler(context.actions.loginUser)}>
                                    <Segment stacked style={{height: '350px'}}>
                                        <Form.Input 
                                            id='username' 
                                            fluid 
                                            icon='user' 
                                            iconPosition='left' 
                                            placeholder='E-mail address'
                                            error={this.state.usernameError} 
                                            onChange={this.changeHandler} 
                                        />
                                        <Form.Input
                                            id='password'
                                            fluid
                                            icon='lock'
                                            iconPosition='left'
                                            placeholder='Password'
                                            type='password'
                                            error={this.state.passwordError} 
                                            onChange={this.changeHandler}
                                        />
                                        {(this.state.usernameError || this.state.passwordError) && this.renderErrorMessage(this.state.errorMessages)}
                                        <Button color='teal' floated='right' size='large'>
                                            Login
                                        </Button>
                                    </Segment>
                                </Form>
                            )}
                        </DataContext.Consumer>
                    </Grid.Column>
                </Grid>
            </div>
        )
    };
};

export default Login;