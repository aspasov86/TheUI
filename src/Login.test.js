import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Login from './Login';
import Overview from './Overview';
import { trueLogin } from './mock-data/constants';

function History() {
    this.history = [];
    this.push = function(a) {
        this.history.push(a);
    }
}

describe('Login Component', () => {
    it('should render without errors', () => {
        expect(shallow(<Login/>).find('div.login-form').exists()).toBe(true);
    })
    it('renders a username input', () => {
        expect(shallow(<Login/>).find('#username').length).toEqual(1);
    })
    
    it('renders a password input', () => {
        expect(shallow(<Login/>).find('#password').length).toEqual(1);
    })
})

describe('Component state change', () => {
    it("should be affected when username is invalid", () => {
        const wrapper = mount(<Login />);
        wrapper.find('input#username').simulate('change', { target: { id: 'username', value: 'aspasov' }});
        wrapper.find('form#submit-btn').simulate('submit');
        expect(wrapper.state('usernameError')).toBe(true);
        expect(wrapper.state('username')).toBe('aspasov');
        expect(wrapper.state('errorMessages')).toContain('Email must be properly formatted');
    })

    it('should be affected when password is invalid', () => {
        const wrapper = mount(<Login/>);
        wrapper.find('form#submit-btn').simulate('submit');
        expect(wrapper.state('passwordError')).toBe(true);
        expect(wrapper.state('password')).toBe('');
        expect(wrapper.state('errorMessages')).toContain('Password must not be empty');
    })
})

describe('User authentication', () => {
    it("should accept correct user's credentials", () => {
        const overview = shallow(<Overview/>);
        const wrapper = mount(<Login history={new History()}/>);
        wrapper.find('input#username').simulate('change', { target: { id: 'username', value: trueLogin.username} });
        wrapper.find('input#password').simulate('change', { target: { id: 'password', value: trueLogin.password} });
        wrapper.find('form#submit-btn').simulate('submit');
        expect(toJson(overview)).toMatchSnapshot();
    })
})