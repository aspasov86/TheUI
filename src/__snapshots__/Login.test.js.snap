// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User authentication should accept correct user's credentials 1`] = `
<div>
  Overview
</div>
`;
